### ISSUE TITLE
Check for missing commas when importing or opening a .bib file

### ISSUE THREAD

**Issue by ThiloteE** [JabRef/jabref#8011]
Not sure if bug or enhancement.

### Problem
Sometimes i forget a comma when editing a bib file.
Sometimes the bib files i get are not formated properly and are missing one or more commas.

When importing such a file in JabRef it leads to a failed import. The error message is very generic and doesn't really help in finding the problem. When opening the file, there is no error, but the library is depicted empty, which does not help finding the problem either.

### What to do to replicate:

- Go to import menu
- filter for bib files in explorer
- import the following bib file:

     
      @inproceedings{kroll2013economics,
        title={The economics of Bitcoin mining, or Bitcoin in the presence of adversaries},
        author={Kroll, Joshua A and Davey, Ian C and Felten, Edward W},
        booktitle={Proceedings of WEIS},
        volume={2013},
        pages={11},
        year={2013}
        url={http://www.infosecon.net/workshop/downloads/2013/pdf/The_Economics_of_Bitcoin_Mining,_or_Bitcoin_in_the_Presence_of_Adversaries.pdf}
      }
      

### Error Message:

Tested with
 
JabRef 5.4--2021-08-21--644e48d
Windows 10 10.0 amd64 
Java 16.0.2 
JavaFX 16+8

- When **importing** the bib file, the following exception stack trace pops up:

        org.jabref.logic.importer.ImportException: Could not find a suitable import format.
	        at org.jabref@5.4.76/org.jabref.logic.importer.ImportFormatReader.importUnknownFormat(Unknown Source)
	        at org.jabref@5.4.76/org.jabref.gui.importer.ImportAction.doImport(Unknown Source)
	        at org.jabref@5.4.76/org.jabref.gui.importer.ImportAction.lambda$automatedImport$1(Unknown Source)
	        at org.jabref@5.4.76/org.jabref.gui.util.BackgroundTask$1.call(Unknown Source)
	        at org.jabref@5.4.76/org.jabref.gui.util.DefaultTaskExecutor$1.call(Unknown Source)
	        at org.jabref.merged.module@5.4.76/javafx.concurrent.Task$TaskCallable.call(Unknown Source)
	        at java.base/java.util.concurrent.FutureTask.run(Unknown Source)
	        at java.base/java.util.concurrent.Executors$RunnableAdapter.call(Unknown Source)
	        at java.base/java.util.concurrent.FutureTask.run(Unknown Source)
	        at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(Unknown Source)
	        at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(Unknown Source)
	        at java.base/java.lang.Thread.run(Unknown Source)

- When **opening** the library normally, there is no error message. It is just shown to be empty. No message, no error, nothing.

### Solution
What JabRef could do to ease the workflow:

- check for missing **,** (comma) in the bib file during the import or when opening a library. (Prior / After an error message?)
- change the error message to a tooltip to the user (with expandable error log) that asks if any commas are missing in the bibfile?
- give an option to automatically insert commas if an error is encountered (not sure if doable?)

### Additional info

The solution could be something similar to what happens when manually editing a file with JabRef in the entry editor.
When deleting a comma of an existing entry in the *{}bibtex source* frame, the following error message will show up:

![image](https://user-images.githubusercontent.com/73715071/130352274-856b9686-f11c-4c99-8ca1-753bb862eea2.png)

I don't really think this is a very nice looking error message, as it makes the user feel as if something terrible just happened and the problem is a VERY SERIOUS CRITICAL ERROR on the part of JabRef, even though it is not JabRef, but the user that might be at fault! At least though, it gives a hint to what happened, where it happened, what the actual problem could be and what JabRef will do.

**The following Error message might be easier to understand:**

- When user input via entry editor in `{}bibtex source` tab:

      User input via entry-editor in `{}bibtex source` tab led to failure.
      Please check your library file for wrong syntax.
   
      Error occured when parsing entry:
      Error in line 11 or above: Empty text token.
      Typical causes:
      - missing comma between two fields.
      - ...
      
      JabRef skipped the entry.
   
- When importing library file:

      Please check your library file for wrong syntax.
      
      Error occured when parsing entry:
      Error in line 11 or above: Empty text token.
      Typical causes:
      - missing comma between two fields.
      - ...
      
      JabRef skipped the entry.
      
- When opening library file:

      Please check your library file for wrong syntax.
      
      Error occured when parsing entry:
      Error in line 11 or above: Empty text token.
      Typical causes:
      - missing comma between two fields.
      - ...
      
      JabRef skipped the entry.

**Comment by ThiloteE** [comment]
IF, and when working on this, in order to be consistent, maybe also other characters apart from the comma could or should be considered checked for.

in addition to the comma error

```
@article{zzz,
title={xyz}
}
```

to account also for the following typo errors:
```
@article{zzz,
title{xyz},
}
```


```
@article{zzz,
={xyz},
}
```

```
@article{zzz,
title={xyz,
}
```
```
@article{zzz,
title=xyz},
}
```
```
@articlezzz,
title={xyz},
}
```
If and how this is feasible from a programming standpoint, i don't know. Just your average user over here 😅

**Comment by dshang1534** [comment]
Hi there,

I'm a graduate student in CS. I have already come up with a solution for this issue. It might not be perfect such as automatically adding commas but it provides a more meaningful prompt as suggested. I attached a screenshot below which shows what the program returns when importing a .bib file with a missing comma (the example from the description).

![Screen Shot 2021-11-19 at 4 54 14 PM](https://user-images.githubusercontent.com/78217710/142696913-58af5c0d-bf1e-4282-9398-b5e0a9c77d31.png)

I plan to continue working on it and provide a few unit tests.

Thanks!

**Comment by Siedlerchr** [comment]
@dshang1534  Thanks for taking a look at this issue. However, I think the dialog could be improved. I would suggest not to show an Exception dialog. That should be really limited to real errors in JabRef. Otherwise, users will think it's a problem of JabRef, but it's just a problem of their library.
Maybe catch the exception from the importer/reader and show a Warning Dialog instead? (have a look at the DialogService class)
And as @ThiloteE  says, it would be cool to have this not only for importing, but also opening. (Two different things).


For creating a PR,  make sure to follow the contribution guide:
Create a PR as soon as possible so we can give you early feedback. 
See here https://devdocs.jabref.org/contributing#understanding-the-basics-of-code-contributions

**Comment by dshang1534** [comment]
@Siedlerchr Thank you for the constructive feedback. I agree with your points. I will try to fix it as much as I can.

**Comment by ThiloteE** [comment]
Awesome, that definitely is one step forward. I wholeheartedly agree with what @Siedlerchr pointed out. Thumps up :)

**Comment by ThiloteE** [comment]
@dshang1534 I am not sure what happened, but if you are still alive and kicking (I hope you are doing fine!), just creating a (draft) pull request with your solution would be nice! Even if you think you will never finish this, just having the possibility to have a look at your proposed solution would help anybody a lot who might think about working on this issue.

**Comment by LIM0000** [comment]
Hi team, I am interested to work with this issue.
Will raise a PR for this issue very soon.

**Comment by ThiloteE** [comment]
Nice :-) Thank you @LIM0000 :)

I was thinking a little bit more about this issue (it has been so long) and edited my first comment. In particular, I added some example error messages that differentiate between "opening a library", "importing a library" and "editing a library via entry-editor".

I was trying to come up with a message that
a) accounts for the fact that JabRef could fail at parsing (--> JabRefs fault)
b) accounts for the fact that users may have library files with faulty sentence structure and syntax
c) users should check their library file for errors and not think that it is JabRefs fault.
d) I now also understand that "empty text token" means that this error message is triggered ALWAYS by a missing comma, but it also could be triggered by other characters. It is just hard for new users to understand what an "empty text token" actually is, so I added an example :)
